megatron/core/inference/engines/dynamic_engine.py

```diff
@@ -358,6 +396,184 @@ async def start_listening_to_data_parallel_coordinator(
+        and resuming the context. TODO(@lmcafee): add argument to optionally
+        return nullcontext, to avoid overhead.
+
+        Args:
+            key (str): Key that identifies caller (e.g., 'suspend' or 'resume').
+
+        Return:
+            None.
+        """
+
+        try:
+
+            start_mem = torch.cuda.memory_stats()
+            start_time = time.time()
+            torch.cuda.synchronize()
+
+            yield
+
+        finally:
+
+            end_time = time.time()
+
+            end_mem = torch.cuda.memory_stats()
+            start_mem_alloc = start_mem["allocated_bytes.all.current"]
+            end_mem_alloc = end_mem["allocated_bytes.all.current"]
+            start_mem_res = start_mem["reserved_bytes.all.current"]
+            end_mem_res = end_mem["reserved_bytes.all.current"]
+
+            rank_str = torch.distributed.get_rank()
+            dir_str = "deallocating" if end_mem_alloc <= start_mem_alloc else "allocating"
+            relative_time_str = f"{end_time - start_time:.3f} sec"
+            relative_mem_str = f"{abs(start_mem_alloc - end_mem_alloc) / 1024**3:.1f} gb"
+
+            process = psutil.Process()
+            mem_info = process.memory_info()
+            total_mem_str = ", ".join((
+                f"cpu: {mem_info.rss / 1024**3:.1f} gb",
+                f"gpu: alloc {end_mem_alloc / 1024**3:.1f} gb",
+                f"res {end_mem_res / 1024**3:.1f} gb",
+            ))
+            logging.info(
+                f"[rank {rank_str}] dynamic engine {key}, "
+                f"unified {unified_memory_level}, "
+                f"{dir_str} "
+                f"{relative_mem_str} in {relative_time_str} ... "
+                f"abs mem usage: {total_mem_str}"
+            )
+
+    def suspend(self):
+        """Suspend engine by deallocating context's GPU state."""
+
+        # Skip if already suspended, which can happen when using the inference
+        # coordinator.
+        if self.is_suspended:
+            return
+        self.is_suspended = True
+
+        # Deallocate context tensors.
+        with self.__class__.suspend_resume_ctx(
+            "suspended", unified_memory_level=self.unified_memory_level
+        ):
+            self.context.deallocate_all_tensors()
+        # >>>
+        print("         SUSPEND request_ids: %s.\n" % str(list(self.requests.keys())))
+        # <<<
+
+        # Delete cuda graphs when not using unified memory at all (level 0). For
+        # levels 1 and 2, the context's tensors maintain static memory addresses,
+        # so the cuda graphs are re-used.
+        if self.unified_memory_level == 0:
+            delete_cuda_graphs()
+
+    def resume(self):
+        """Resume engine by reallocating context's GPU state."""
+
+        # Skip if not suspended, which can happen when using the inference
+        # coordinator.
+        if not self.is_suspended:
+            return
+        self.is_suspended = False
```

Comment:
Put a check in `async_step` to prevent stepping while the engine is suspended. And make the condition on L1164 inside `run_engine` also check for `not self.is_suspended`, then run `_notify_cond_for_new_request` (examples elsewhere in this file) upon request to wake up the `run_engine` condition.